xds/src/main/java/io/grpc/xds/RbacFilter.java

```diff
@@ -0,0 +1,325 @@
+import io.grpc.Status;
+import io.grpc.xds.Filter.ServerInterceptorBuilder;
+import io.grpc.xds.internal.MatcherParser;
+import io.grpc.xds.internal.Matchers;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.AlwaysTrueMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.AndMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.AuthConfig;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.AuthDecision;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.AuthHeaderMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.AuthenticatedMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.DestinationIpMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.DestinationPortMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.InvertMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.Matcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.OrMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.PathMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.PolicyMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.RequestedServerNameMatcher;
+import io.grpc.xds.internal.rbac.engine.GrpcAuthorizationEngine.SourceIpMatcher;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/** RBAC Http filter implementation. */
+final class RbacFilter implements Filter, ServerInterceptorBuilder {
+  private static final Logger logger = Logger.getLogger(RbacFilter.class.getName());
+
+  static final RbacFilter INSTANCE = new RbacFilter();
+
+  static final String TYPE_URL =
+          "type.googleapis.com/envoy.extensions.filters.http.rbac.v3.RBAC";
+
+  private static final ServerCall.Listener<?> NOOP_LISTENER = new NoopListener<>();
+
+  RbacFilter() {}
+
+  @Override
+  public String[] typeUrls() {
+    return new String[] { TYPE_URL };
+  }
+
+  @Override
+  public ConfigOrError<RbacConfig> parseFilterConfig(Message rawProtoMessage) {
+    RBAC rbacProto;
+    if (!(rawProtoMessage instanceof Any)) {
+      return ConfigOrError.fromError("Invalid config type: " + rawProtoMessage.getClass());
+    }
+    Any anyMessage = (Any) rawProtoMessage;
+    try {
+      rbacProto = anyMessage.unpack(RBAC.class);
+    } catch (InvalidProtocolBufferException e) {
+      return ConfigOrError.fromError("Invalid proto: " + e);
+    }
+    return parseRbacConfig(rbacProto);
+  }
+
+  @VisibleForTesting
+  static ConfigOrError<RbacConfig> parseRbacConfig(RBAC rbac) {
+    checkNotNull(rbac, "rbac");
+    if (!rbac.hasRules()) {
+      return null;
```

Comment:
The `Filter.parseFilterConfig()` API was not returning `@Nullable` type, and currently all its usage is not handling null return value. It's possible NPE is thrown in the runtime.

Reply:
Did you mean `return RbacConfig.create(null)`?

Reply:
Yes, I checked code and you that's the case, I might have been misunderstood with https://github.com/grpc/grpc-java/blob/b7f3fddc76ecfe854c6039c000ff6d168205ab06/xds/src/main/java/io/grpc/xds/ClientXdsClient.java#L324
It looks we should not return null for now.
However, wondering how do we now we distinguish ignored or error, should't we? Shouldn't xdsClient behave differently?

Reply:
Currently we use `null` if we want ClientXdsClient to ignore the config. But currently for the `Filter.parseFilterConfig()` API itself, it should always return a config or an error, because it does not know whether `ClientXdsClient` will ignore it.

Reply:
> Did you mean `return ConfigOrError.fromConfig(RbacConfig.create(null))`?

No, I think this is `ConfigOrError.fromConfig(RbacConfig.create(null))` seems fine as a way to disable global rbac filter.

I was asking that does the place that uses `Filter.parseFilterConfig()` need to distinguish between a ignored `FilterConfig` (e.g. if there is `LOG` action in a rbacConfig then this config should be ignored, as if not received. )and an error `FilterConfig`